app/abuser_utils.py

```diff
@@ -0,0 +1,197 @@
+    """
+    Returns False, if the given address (after hashing) is found in abuser_lookup.
+    """
+    mac_key_bytes = config.MAC_KEY
+    check_hmac = hmac.new(
+        mac_key_bytes, new_address.encode("utf-8"), sha256
+    ).hexdigest()
+
+    return (
+        AbuserLookup.filter(AbuserLookup.hashed_address == check_hmac).limit(1).first()
+    )
+
+
+def archive_abusive_user(user: User) -> None:
+    """
+    Archive the given abusive user's data and update blocklist/lookup tables.
+    """
+    if not user.email:
+        raise ValueError(f"User ID {user.id} must have a primary email to be archived.")
+
+    try:
+        primary_email: str = user.email
+        aliases: List[Alias] = Alias.filter_by(user_id=user.id).all()
+        mailboxes: List[Mailbox] = Mailbox.filter_by(user_id=user.id).all()
+        bundle = {
+            "account_id": user.id,
+            "email": primary_email,
+            "user_created_at": user.created_at.isoformat() if user.created_at else None,
+            "aliases": [
+                {
+                    "address": alias.email,
+                    "created_at": alias.created_at.isoformat()
+                    if alias.created_at
+                    else None,
+                }
+                for alias in aliases
+            ],
+            "mailboxes": [
+                {
+                    "address": mailbox.email,
+                    "created_at": mailbox.created_at.isoformat()
+                    if mailbox.created_at
+                    else None,
+                }
+                for mailbox in mailboxes
+            ],
+        }
+        bundle_json_bytes = json.dumps(bundle, sort_keys=True).encode("utf-8")
+        derived_enc_key = _derive_encryption_key(config.MASTER_ENC_KEY, user.id)
+        aesgcm = AESGCM(derived_enc_key)
+        nonce = secrets.token_bytes(12)
+        aad_bytes = config.AEAD_AAD_DATA.encode("utf-8")
+        encrypted_bundle_data = nonce + aesgcm.encrypt(
+            nonce, bundle_json_bytes, aad_bytes
+        )
+        abuser_data_entry = AbuserData(
+            user_id=user.id, encrypted_bundle=encrypted_bundle_data
+        )
+
+        Session.add(abuser_data_entry)
+        Session.flush()
+
+        blob_id = abuser_data_entry.id
+        all_identifiers = (
+            [primary_email] + [a.email for a in aliases] + [m.email for m in mailboxes]
+        )
+        seen_identifiers = set()
+        mac_key_bytes = config.MAC_KEY
+
+        for identifier_address in all_identifiers:
+            if not identifier_address or identifier_address in seen_identifiers:
+                continue
+
+            seen_identifiers.add(identifier_address)
+
+            identifier_hmac = hmac.new(
+                mac_key_bytes, identifier_address.lower().encode("utf-8"), sha256
+            ).hexdigest()
+            abuser_lookup_entry = AbuserLookup(
+                hashed_address=identifier_hmac, abuser_data_id=blob_id
+            )
+
```

Comment:
We're missing a `user.disabled=True` to disable the user properly

Reply:
Shouldn't the disabling/enabling switch happen in the upstream flow that actually responsible for marking the user as an abuser? To me, the abuser_utils looks more like a utility side flow to just store the encrypted bundle.